### PR TITLE
Souls

### PR DESCRIPTION
**Pull request by aristotlepenguin**
-Every player gets 1 soul.
-You can !extractsoul to remove your soul from your body. You can exchange and bet this soul as though it was any other cosmetic item.
-You can regain a soul by using !returnsoul with a soul in your inventory. The soul needs to be yours for it to work. 
-If you hold a soul from another player, you can !squeezesoul <player> <message> to taunt them, and drain them of a big chunk of slime. Threaten using this power to make them do your bidding! You also can send a message, similar to haunt messaging.
-You can !betsoul in the casino to exchange it for half a billion Slimecoin. However, you and other players can !buysoul to purchase a soul that has been gambled this way. Mention a player when betting or buying to specify their soul.

-Also, I added picture frames. You can !frame an image link to make it a part of the item description.

NOTE: you'll probably see it but there's a SQL field you have to add

### CODE REVIEW

File: ewcasino.py [hunk 1]
```diff
@@ -2022,3 +2022,47 @@ def determine_trick_taker(trick,gametype,trump):
...
+
+async def buysoul(cmd):
+	usermodel = EwUser(id_user=cmd.message.author.id, id_server=cmd.message.server.id)
+	casino_inv = ewitem.inventory(id_user="casinosouls", id_server=cmd.message.server.id, item_type_filter=ewcfg.it_cosmetic)
+	selected_item = None
+	for item_sought in casino_inv:
+		item_object = ewitem.EwItem(item_sought.get('id_item'))
+		selected_item = item_object
+		if item_object.item_props.get('user_id') == cmd.message.author.id:
+			break
```
Comment (magicksRampage) [inline ewcasino.py]: you should be able to buy a specific person's soul

File: ewcasino.py [hunk 2]
```diff
@@ -2022,3 +2022,47 @@ def determine_trick_taker(trick,gametype,trump):
...
+
+async def betsoul(cmd):
+	usermodel = EwUser(id_user=cmd.message.author.id, id_server=cmd.message.server.id)
+	user_inv = ewitem.inventory(id_user=cmd.message.author.id, id_server=cmd.message.server.id, item_type_filter=ewcfg.it_cosmetic)
+	item_select = None
+	for item in user_inv:
+		item_object = ewitem.EwItem(item.get('id_item'))
+		if item_object.item_props.get('id_cosmetic') == "soul":
+			item_select = item_object
+			break
```
Comment (magicksRampage) [inline ewcasino.py]: you should be able to bet a specific soul

File: ewcasino.py [hunk 3]
```diff
@@ -2022,3 +2022,47 @@ def determine_trick_taker(trick,gametype,trump):
...
+
+	if usermodel.poi != ewcfg.poi_id_thecasino:
+		response = "If you want to exchange your soul for SlimeCoin you have to be in the casino first."
+	elif item_select == None:
+		response = "You don't have any souls in your inventory. !extractsoul if you want to do this properly."
+	else:
+		ewitem.give_item(id_user="casinosouls", id_server=cmd.message.server.id, id_item=item_select.id_item)
+		usermodel.change_slimecoin(coinsource=ewcfg.coinsource_spending, n=ewcfg.soulprice) #current price for souls is 500 mil slimecoin
+		usermodel.persist()
+		response = "You hand over {} for 500 million slimecoin.".format(item_select.item_props.get('cosmetic_name'))
```
Comment (magicksRampage) [inline ewcasino.py]: I suggest using the soulprice variable here

File: ewitem.py [hunk 4]
```diff
@@ -1396,6 +1398,115 @@ def gen_item_props(item):
...
+				'cosmetic_name': "{}'s soul".format(playermodel.display_name),
+				'cosmetic_desc': "The immortal soul of {}. It dances with a vivacious energy inside its jar.\n If you listen to it closely you can hear it whispering numbers: {}.".format(playermodel.display_name, cmd.message.author.id),
+				'rarity': ewcfg.rarity_patrician,
+				'adorned': 'false',
+				'user_id': usermodel.id_user
+			}
+		)
+		usermodel.has_soul = 0
+		usermodel.persist()
+		response = "You tremble at the thought of trying this. Nothing ventured, nothing gained, you suppose. With all your mental fortitude you jam your hand deep into your chest and begin to pull out the very essence of your being. Your spirit, aspirations, everything that made you who you are begins to slowly drain from your mortal effigy until you feel absolutely nothing. Your soul flickers about, taunting you from outside your body. You capture it in a jar, almost reflexively.\n\nWow. You personality must suck now."
```
Comment (magicksRampage) [inline ewitem.py]: typo "Your personality"

File: ewitem.py [hunk 5]
```diff
@@ -1396,6 +1398,115 @@ def gen_item_props(item):
...
+	else:
+		response = "There's nothing left in you to extract. You already spent the soul you had."
+	return await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(cmd.message.author, response))
+		
+async def returnsoul(cmd):
+	usermodel = EwUser(member=cmd.message.author)
+	soul = find_item(item_search="soul", id_user=cmd.message.author.id, id_server=cmd.message.server.id)
```
Comment (magicksRampage) [inline ewitem.py]: you might not be able to return your own soul, if you have another person's soul in your inventory and the search returns that one first

File: ewitem.py [hunk 6]
```diff
@@ -1396,6 +1398,115 @@ def gen_item_props(item):
...
+			poi = ewcfg.id_to_poi.get(targetmodel.poi)
+
+			usermodel.time_lasthaunt = int(time.time())
+			usermodel.persist()
+
+			server = ewcfg.server_list[targetmodel.id_server]
+			member_object = server.get_member(targetmodel.id_user)
+
+			penalty = (targetmodel.slimes* -0.25)
+			targetmodel.change_slimes(n=penalty, source=ewcfg.source_haunted)
```
Comment (magicksRampage) [inline ewitem.py:1508]: is it supposed to work on ghosts? cause currently it does

**Review by magicksRampage — CHANGES_REQUESTED** [review]

File: ewitem.py (hunk 6 above)
Comment (aristotlepenguin) [reply ewitem.py:1508]: it was, but it might be better if it didn't.

**Pull request merged**